### PR TITLE
Reduce comparator objects init cost in BlockIter

### PR DESCRIPTION
**Pull request by XinyuZeng**
This PR solves the problem discussed in #7149. By storing the pointer of InternalKeyComparator as icmp_ in BlockIter, the object size remains the same. And for each call to CompareCurrentKey, there is no need to create Comparator objects. One can use icmp_ directly or use the "user_comparator" from the icmp_.

Test Plan: with #9903,

```
$ TEST_TMPDIR=/dev/shm python3.6 ../benchmark/tools/compare.py benchmarks ./db_basic_bench ../rocksdb-pr9611/db_basic_bench --benchmark_filter=DBGet/comp_style:0/max_data:134217728/per_key_size:256/enable_statistics:0/negative_query:0/enable_filter:0/mmap:1/iterations:262144/threads:1 --benchmark_repetitions=50
...
Comparing ./db_basic_bench to ../rocksdb-pr9611/db_basic_bench                                                                                                                                                                                            
Benchmark                                                                                                                                                               Time             CPU      Time Old      Time New       CPU Old       CPU New
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
...
DBGet/comp_style:0/max_data:134217728/per_key_size:256/enable_statistics:0/negative_query:0/enable_filter:0/mmap:1/iterations:262144/threads:1_pvalue                 0.0001          0.0001      U Test, Repetitions: 50 vs 50
DBGet/comp_style:0/max_data:134217728/per_key_size:256/enable_statistics:0/negative_query:0/enable_filter:0/mmap:1/iterations:262144/threads:1_mean                  -0.0483         -0.0483          3924          3734          3924          3734
DBGet/comp_style:0/max_data:134217728/per_key_size:256/enable_statistics:0/negative_query:0/enable_filter:0/mmap:1/iterations:262144/threads:1_median                -0.0713         -0.0713          3971          3687          3970          3687
DBGet/comp_style:0/max_data:134217728/per_key_size:256/enable_statistics:0/negative_query:0/enable_filter:0/mmap:1/iterations:262144/threads:1_stddev                -0.0342         -0.0344           225           217           225           217
DBGet/comp_style:0/max_data:134217728/per_key_size:256/enable_statistics:0/negative_query:0/enable_filter:0/mmap:1/iterations:262144/threads:1_cv                    +0.0148         +0.0146             0             0             0             0
OVERALL_GEOMEAN                                                                                                                                                      -0.0483         -0.0483             0             0             0             0
```

### CODE REVIEW

**Review by ajkr — APPROVED** [review]
LGTM. Updated the "Test Plan" to show results of microbenchmark

File: table/block_based/block.h [hunk 1]
```diff
@@ -405,11 +390,23 @@ class BlockIter : public InternalIteratorBase<TValue> {
...
-  }
+  void InitializeBase(const Comparator* raw_ucmp, const char* data,
+                      uint32_t restarts, uint32_t num_restarts,
+                      SequenceNumber global_seqno, bool block_contents_pinned) {
+    assert(data_ == nullptr);  // Ensure it is called only once
+    assert(num_restarts > 0);  // Ensure the param is valid
 
-  UserComparatorWrapper ucmp() { return UserComparatorWrapper(raw_ucmp_); }
+    icmp_ =
+        std::make_unique<InternalKeyComparator>(raw_ucmp, false /* named */);
```
Comment (ajkr) [inline table/block_based/block.h:400]: Can `InternalKeyComparator` wrap a `UserComparatorWrapper`?

Comment (XinyuZeng) [reply table/block_based/block.h:400]: It seems like InternalKeyComparator already has a UserComparatorWrapper as a private member. Calling icmp_->user_comparator() refers to it.

**Comment by XinyuZeng** [comment]
hey @ajkr, do you mind share the compare.py script you used in the Test Plan? Thanks!

**Comment by ajkr** [comment]
> hey @ajkr, do you mind share the compare.py script you used in the Test Plan? Thanks!

Sure, it's a script provided in google benchmark -- https://github.com/google/benchmark/blob/main/tools/compare.py

**Comment by riversand963** [comment]
@XinyuZeng can you take a look at #10340 ?

**Comment by XinyuZeng** [comment]
@riversand963 Through a quick look, I am not sure why #9611 introduces more allocations using glibc because it should reduce the number of allocations compared to the commit before it. Before the commit, each call to CompareCurrentKey will have an allocation. But I do agree with #10342:
> As a side effect, internal key comparator was made configurable too. This introduces overhead to this simple wrapper. For example, every InternalKeyComparator will have an std::vector attached to it, which consumes memory and possible allocation overhead too.

This is also what I profiled and found when I did #9611

**Pull request closed without merging**